### PR TITLE
Fix tangential polygons

### LINKED ISSUE #19471: Invalid results by `is_tangent`
What's Wrong:
```python
>>> from sympy import *
>>> c = Circle(Point(0,0),2)
>>> poly1 = Polygon((-2,-2),(-2,2),(2,2),(3,0),(2,-2))
>>> c.is_tangent(poly1)
True
>>> poly2 = Polygon((2,2),(3,0),(2,-2),(-2,-2),(-2,2))
>>> c.is_tangent(poly2)
False
>>> r = Ray(Point(0,2),Point(2,2))
>>> c.is_tangent(r)
False
```

notice that poly1 and poly2 are giving different results,
but are same polygon except there starting points are different.
Further, in my opinion( I found very less on this topic on internet but by definition in google),  tangent is a straight line or plane that touches a curve or curved surface at a point, but if EXTENDED does not cross it at that point.
the ray,r touches the circle,c at r.source but when extended does not cross circle,c at any point. by this definition r is a tangent to circle c.

Changes I made:
change 1-
Line Number:720-730
       if one or more sides of polygon are tangent to ellipse and other does not touch it, polygon is a tangent
      

change 2-
Line Number: 714
       original content : 
                return intersect[0] != o.source and not self.encloses_point(o.source)
       my change:
                return o in self.tangent_lines(intersect[0])[0]

Orignally reported at https://github.com/sympy/sympy/pull/13496 by @Amit-Modi

### PR DESCRIPTION
**Pull request by Stam21**
#### References to other Issues or PRs
Fixes #19471 

#### Brief description of what is fixed or changed
All sides of the polygon must be tangent to the circle.  After finding a valid intersection, there was a case that could return true even though the next segment has no intersection.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by sylee957** [comment]
I think that you have implement something other than the original author of #13496
Because it now uses the definition to assume *all* vertices of the polygon to be tangent to the circle than *any* vertices of the polygon to be tangent to the circle.

I think that we need some agreement onto define how `is_tangent` should work between polygon and circle.
Have you addressed https://github.com/sympy/sympy/pull/13496#issuecomment-344555246?
And have you also read #15774?

**Comment by Stam21** [comment]
Basically, a tangent polygon to a circle can be translated into circumscribed polygon of a circle. That means that every line that is formed between two consecutive points has to be tangent to the circle. I struggle to find anything different

File: sympy/geometry/ellipse.py [hunk 1]
```diff
@@ -762,7 +762,7 @@ def is_tangent(self, o):
                     else:
                         return False
                 else:
-                    return all_tangents
+                    return False
```
Comment (sylee957) [inline sympy/geometry/ellipse.py:765]:
```suggestion
                    return False
```

**Comment by sylee957** [comment]
I agree that there are more interesting properties for polygons of all vertices tangent to a circle than any vertices tangent to a circle.
But I see that the original issue is also about the intersection between ray and the circle, are you going to fix the issue if you mark that as `Fixed`?

And you should add the test cases for this.

**Comment by Stam21** [comment]
A ray is tangent when there is an intersection between the starting point
and infinite towards the second point. When the starting point intersects
with circle I don't think it should be considered tangent, as sympy
correctly outputs. But if you think otherwise I can work on that and finish
the pull request of course. Thanks for reaching out!

**Comment by sylee957** [comment]
Okay I'd defer how to handle is_tangent for Ray.

**Comment by Stam21** [comment]
So should I just add test cases for the issue with the polygons?

**Comment by sylee957** [comment]
Yes, and you should make sure the travis test passes

**Comment by oscarbenjamin** [comment]
The Travis test has failed because of trailing whitespace:
```
AssertionError: File contains trailing whitespace:
 /home/travis/build/sympy/sympy/sympy/geometry/ellipse.py, line 765.
```

**Comment by eric-wieser** [comment]
Fixed the whitespace, but this still needs a test

**Comment by czgdp1807** [comment]
@Stam21 Are you still working on this? Please do not close this PR. Thanks for your contributions.

**Pull request closed without merging**